src/utils/helpers/GraphHelpers.ts

```diff
@@ -1,187 +1,298 @@
+        } else {
+            select<Element, DependencyNode>(labelElement).attr('fill', Colors.DEFAULT);
+            select<Element, DependencyNode>(textElement).style('fill', 'black');
+        }
+    });
 }
 
-export function generateMarkers(svgContainer: GraphContainer): void {
+export function selectHighLightedNodes() {
+    return selectAllNodes().filter(function(this: SVGGElement) {
+        return this.firstElementChild ? this.firstElementChild.getAttribute('fill') !== Colors.DEFAULT : false;
+    });
+}
+
+export function selectAllNodes() {
+    return select('#labels').selectAll<SVGGElement, DependencyNode>('g');
+}
+
+export function zoomToHighLightedNodes() {
+    const highlightedNodes = selectHighLightedNodes();
+    const svgContainer = select('#container');
+    const svgContainerNode = select<SVGSVGElement, DependencyNode>('#container').node();
+    const dim = findGroupBackgroundDimension(highlightedNodes.data());
+
+    if (!svgContainerNode || !dim) {
+        return;
+    }
+
+    const width = Number(svgContainerNode.getAttribute('width'));
+    const height = Number(svgContainerNode.getAttribute('height'));
+
+    const scaleValue = Math.min(8, 0.9 / Math.max(dim.width / width, dim.height / height));
+
     svgContainer
-        .append('svg:defs')
-        .append('svg:marker')
-        .attr('id', 'provider')
-        .attr('viewBox', '-5 -5 40 10')
-        .attr('refX', 15)
-        .attr('refY', 0)
-        .attr('markerWidth', 40)
-        .attr('markerHeight', 40)
-        .attr('orient', 'auto')
-        .append('svg:path')
-        .attr('d', 'M0,-5L20,0L0,5,q10 -5,0 -10')
-        .attr('fill', '#E5E5E6');
-}
-
-export function generateLabelPath(this: Node, node: DependencyNode) {
-    const labelTextDimensions = getLabelTextDimensions(this);
-
-    if (!labelTextDimensions) {
-        return '';
+        .transition()
+        .duration(750)
+        .call(
+            zoom<any, any>().on('zoom', zoomed).transform,
+            zoomIdentity
+                .translate(width / 2, height / 2)
+                .scale(scaleValue)
+                .translate(-dim.x - dim.width / 2 - 250, -dim.y - dim.height / 2)
+        );
+}
+
+export function setDependencyLevelOnEachNode(clickedNode: DependencyNode, nodes: DependencyNode[]): DependencyNode[] {
+    nodes.forEach((node: DependencyNode) => (node.level = 0));
+
+    const visitedNodes: DependencyNode[] = [];
+    const nodesToVisit: DependencyNode[] = [];
+
+    nodesToVisit.push({ ...clickedNode, level: 1 });
+
+    while (nodesToVisit.length > 0) {
+        const currentNode = nodesToVisit.shift();
+
+        if (!currentNode) {
+            return [];
+        }
+
+        currentNode.links.forEach((node: DependencyNode) => {
+            if (!contains(visitedNodes, node) && !contains(nodesToVisit, node)) {
+                node.level = currentNode.level + 1;
+                nodesToVisit.push({ ...node, level: currentNode.level + 1 });
```

Comment:
why do you desctructure ``node``  object ?

why do you not write just like that:
```typescript
                node.level = currentNode.level + 1;
                nodesToVisit.push(node);
```

and the same with ``clickedNode`` ?
```typescript
clickedNode.level = 1;
nodesToVisit.push(clickedNode);
```

and then you could write ``contains`` function as:
```typescript
function contains(arr: DependencyNode[], node: DependencyNode) {
    return Boolean(arr.find((el: DependencyNode) =>el === node));
}
```

Reply:
The first part about destructuring is leftover after code refactor, gonna change that. About contains function, if not necessary I would rather not compare with reference, also findIndex unlike find is supported by IE11 :))

Reply:
Array.find is probably polyfilled by Babel.

IT specialists who will use **Judge Dreed UI**, do not use IE-11 :D